### ISSUE TITLE
Is `:=` supposed to mutate columns on other data.tables?

### ISSUE THREAD

**Issue by nhirschey** [Rdatatable/data.table#4310]
I understand that := mutates columns in place. But it will also change other data.tables. Is this intended? It seems very easy to introduce bugs this way (in fact I did).  

In the below example, I do not expect that adding a column to data.table `y` will also add a column to data.table `x`.

The problem is that `<-` is not creating a new data.table, but appears to instead reference the original.

```r
> x <- data.table(a=1:4)
> y <- x
> y[,b3:=3]
> x
   a b3
1: 1  3
2: 2  3
3: 3  3
4: 4  3
```
`#` `Output of sessionInfo()`
```r
> sessionInfo()
R version 3.6.2 (2019-12-12)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19041)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                           LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] data.table_1.12.9

loaded via a namespace (and not attached):
[1] compiler_3.6.2 tools_3.6.2   
```

**Comment by nhirschey** [comment]
I have now found the `copy` documentation (I did look before posting this issue, sorry for not finding it). The copy documentation does reveal that this behavior is intended.

This feels dangerous that `y <- copy(x)` is needed to get the standard R assignment behavior of creating a wholly different copy. I am a novice data.table user, but I know R well. And this is strikes me as a dangerous default for `<-`.

**Comment by tdeenes** [comment]
Did you have a close look at `?data.table::set` ?

Note that `<-` in base R does not create a copy, either. E.g.:

```r
library(data.table)
x <- 1:3
ax <- address(x) # in-memory address of 'x'
y <- x
ay <- address(y) # in-memory address of 'y'
identical(ax, ay) # TRUE
```

It is a non-trivial issue how R "realizes" when a copy of 'x' or 'y' is needed (see references).

The set* operators and functions in data.table must be used with extreme care, but if done so, they can result in a **huge** gain in performance (both in terms of speed and memory efficiency). 

Some references:
https://stackoverflow.com/questions/10225098/understanding-exactly-when-a-data-table-is-a-reference-to-vs-a-copy-of-another?rq=1
https://adv-r.hadley.nz/names-values.html
https://www.brodieg.com/2019/02/18/an-unofficial-reference-for-internal-inspect/
https://developer.r-project.org/Refcnt.html

**Comment by MichaelChirico** [comment]
With great power comes great responsibility -- please check the [Reference semantics](https://cran.r-project.org/web/packages/data.table/vignettes/datatable-reference-semantics.html) vignette. 
data.table is designed with massive in-memory tables (e.g. 1 billion rows) in mind.

In such scenarios, diligently avoiding unnecessary copies is extremely important to efficient processing.

We are quite open to how we could improve our documentation for this to have come up in your original search, please re-open if there is a gap you would like to help close

**Comment by nhirschey** [comment]
Thanks to both of you for the helpful references. I appreciate the utility of avoiding copies for speed with big datasets, which is why I decided to finally figure out how to use data.table properly. 

The existing documentation is extensive and helpful, but I think this issue with `:=` and `<-` could be more clear. I had read the reference semantics vignette, with a particular eye to issues related to unintentional mutation, but I did not appreciate the relevance of the bottom section discussing `copy`. I will do a documentation pull request with some suggestions to highlight this behavior more clearly.
